In [1]:
# install main library YFinance
!pip install yfinance

     ---------------------------------------- 0.0/3.0 MB ? eta -:--:--
     -- ------------------------------------- 0.2/3.0 MB 5.9 MB/s eta 0:00:01
     ------ --------------------------------- 0.5/3.0 MB 5.6 MB/s eta 0:00:01
     --------- ------------------------------ 0.7/3.0 MB 6.2 MB/s eta 0:00:01
     --------- ------------------------------ 0.7/3.0 MB 4.5 MB/s eta 0:00:01
     ----------------- ---------------------- 1.3/3.0 MB 5.7 MB/s eta 0:00:01
     -------------------- ------------------- 1.5/3.0 MB 6.0 MB/s eta 0:00:01
     ----------------------- ---------------- 1.8/3.0 MB 5.6 MB/s eta 0:00:01
     ---------------------------- ----------- 2.1/3.0 MB 6.0 MB/s eta 0:00:01
     ------------------------------ --------- 2.2/3.0 MB 5.5 MB/s eta 0:00:01
     --------------------------------- ------ 2.4/3.0 MB 5.4 MB/s eta 0:00:01
     ------------------------------------- -- 2.8/3.0 MB 5.7 MB/s eta 0:00:01
     ---------------------------------------- 3.0/3.0 MB 5.5 MB/s eta 0

In [3]:
# install main library pandas-datareader,plotly, matplotlib
!pip install pandas-datareader plotly matplotlib

  Using cached tenacity-8.2.3-py3-none-any.whl.metadata (1.0 kB)
     ---------------------------------------- 0.0/162.8 kB ? eta -:--:--
     ------------------------------- ------ 133.1/162.8 kB 2.6 MB/s eta 0:00:01
     -------------------------------------- 162.8/162.8 kB 3.2 MB/s eta 0:00:00
   ---------------------------------------- 0.0/109.5 kB ? eta -:--:--
   ---------------------------------------- 109.5/109.5 kB ? eta 0:00:00
   ---------------------------------------- 0.0/15.7 MB ? eta -:--:--
    --------------------------------------- 0.2/15.7 MB 4.7 MB/s eta 0:00:04
    --------------------------------------- 0.4/15.7 MB 7.8 MB/s eta 0:00:02
    --------------------------------------- 0.4/15.7 MB 7.8 MB/s eta 0:00:02
   -- ------------------------------------- 0.9/15.7 MB 5.3 MB/s eta 0:00:03
   --- ------------------------------------ 1.2/15.7 MB 5.9 MB/s eta 0:00:03
   --- ------------------------------------ 1.5/15.7 MB 5.7 MB/s eta 0:00:03
   ---- ------------------

In [34]:
# IMPORTS
import numpy as np
import pandas as pd

#Fin Data Sources
import yfinance as yf
import pandas_datareader as pdr

#Data viz
import plotly.graph_objs as go
import plotly.express as px

import time
from datetime import date
from datetime import datetime

Question 1: [Macro] Average growth of GDP in 2023
What is the average growth (in %) of GDP in 2023?

In [45]:
start = date(year=2022, month=1, day=1)
df = pdr.DataReader("GDPC1", "fred", start=start)
df.reset_index(inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   DATE    8 non-null      datetime64[ns]
 1   GDPC1   8 non-null      float64       
dtypes: datetime64[ns](1), float64(1)
memory usage: 256.0 bytes


In [46]:
df['gdp_growth_yoy'] = df.GDPC1/df.GDPC1.shift(4)-1
df_2023 = df[df['DATE'].dt.year == 2023]
average_gdpc1_us_yoy_2023 = round(df_2023['gdp_growth_yoy'].mean() * 100, 1)
print(f"Average Year to Year GDP Growth in 2023: {average_gdpc1_us_yoy_2023}%")

Average Year to Year GDP Growth in 2023: 2.5%


In [86]:
import pandas as pd
import pandas_datareader as pdr

# Download GDPC1 data from FRED
gdpc1_data = pdr.get_data_fred('GDPC1')

# Calculate year-over-year (YoY) growth rate
gdpc1_data['YoY_growth'] = gdpc1_data['GDPC1'].pct_change(4) * 100

# Filter data for 2023
gdpc1_data_2023 = gdpc1_data[gdpc1_data.index.year == 2023]

# Calculate the average YoY growth rate for 2023
average_growth_2023 = gdpc1_data_2023['YoY_growth'].mean()

# Round the result to one decimal point
average_growth_2023 = round(average_growth_2023, 1)

print("Average GDP YoY growth rate in 2023: {}%".format(average_growth_2023))


Average GDP YoY growth rate in 2023: 2.5%


Question 2. [Macro] Inverse "Treasury Yield"

Find the min value of (dgs10-dgs2) after since year 2000 (2000-01-01) and write it down as an answer, round to 1 digit after the decimal point.

Download DGS2 and DGS10 interest rates series (https://fred.stlouisfed.org/series/DGS2, https://fred.stlouisfed.org/series/DGS10). 
Join them together to one dataframe on date (you might need to read about pandas.DataFrame.join()), calculate the difference dgs10-dgs2 daily.

(Additional: think about what does the "inverted yield curve" mean for the market and investors? 
do you see the same thing in your country/market of interest? Do you think it can be a good predictive feature for the models?)

In [52]:
df_dgs2 = pdr.DataReader("DGS2", "fred", start=date(year=2000, month=1, day=1))
print(df_dgs2)
df_dgs10 = pdr.DataReader("DGS10", "fred", start=date(year=2000, month=1, day=1))
print(df_dgs10)

            DGS2
DATE            
2000-01-03  6.38
2000-01-04  6.30
2000-01-05  6.38
2000-01-06  6.35
2000-01-07  6.31
...          ...
2024-04-12  4.88
2024-04-15  4.93
2024-04-16  4.97
2024-04-17  4.93
2024-04-18  4.98

[6339 rows x 1 columns]
            DGS10
DATE             
2000-01-03   6.58
2000-01-04   6.49
2000-01-05   6.62
2000-01-06   6.57
2000-01-07   6.52
...           ...
2024-04-12   4.50
2024-04-15   4.63
2024-04-16   4.67
2024-04-17   4.59
2024-04-18   4.64

[6339 rows x 1 columns]


In [54]:
df_merge = df_dgs2.join(df_dgs10)
df_merge['dgs10-dgs2'] = df_merge['DGS10']-df_merge['DGS2']
print(df_merge)
min_dgs = round(df_merge['dgs10-dgs2'].min(), 1)
print('Min value of dgs', min_dgs)

            DGS2  DGS10  dgs10-dgs2
DATE                               
2000-01-03  6.38   6.58        0.20
2000-01-04  6.30   6.49        0.19
2000-01-05  6.38   6.62        0.24
2000-01-06  6.35   6.57        0.22
2000-01-07  6.31   6.52        0.21
...          ...    ...         ...
2024-04-12  4.88   4.50       -0.38
2024-04-15  4.93   4.63       -0.30
2024-04-16  4.97   4.67       -0.30
2024-04-17  4.93   4.59       -0.34
2024-04-18  4.98   4.64       -0.34

[6339 rows x 3 columns]
Min value of dgs -1.1


Question 3. [Index] Which Index is better recently?

Compare S&P 500 and IPC Mexico indexes by the 5 year growth and write down the largest value as an answer (%)

Download on Yahoo Finance two daily index prices for S&P 500 (^GSPC, https://finance.yahoo.com/quote/%5EGSPC/) and 
IPC Mexico (^MXX, https://finance.yahoo.com/quote/%5EMXX/). Compare 5Y growth for both (between 2019-04-09 and 2024-04-09). 
Select the higher growing index and write down the growth in % (closest integer %). 
E.g. if ratio end/start was 2.0925 (or growth of 109.25%), you need to write down 109 as your answer.

(Additional: think of other indexes and try to download stats and compare the growth? 
Do create 10Y and 20Y growth stats. What is an average yearly growth rate (CAGR) for each of the indexes you select?)

In [55]:
import yfinance as yf
import numpy as np

# Download historical data for S&P 500 (^GSPC)
sp500_data = yf.download('^GSPC', start='2019-04-09', end='2024-04-09')['Adj Close']

# Download historical data for IPC Mexico (^MXX)
ipc_data = yf.download('^MXX', start='2019-04-09', end='2024-04-09')['Adj Close']

# Calculate 5-year growth rates for both indexes
sp500_growth = (sp500_data.iloc[-1] / sp500_data.iloc[0] - 1) * 100
ipc_growth = (ipc_data.iloc[-1] / ipc_data.iloc[0] - 1) * 100

# Determine which index has the higher growth rate
better_index = 'S&P 500' if sp500_growth > ipc_growth else 'IPC Mexico'
largest_growth = np.round(max(sp500_growth, ipc_growth))

print("Better index recently: {}".format(better_index))
print("5-year growth rate: {}%".format(int(largest_growth)))

[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed

Better index recently: S&P 500
5-year growth rate: 81%


In [62]:
import yfinance as yf

# Download S&P 500 (^GSPC) data from Yahoo Finance
sp500_data = yf.download('^GSPC', start='2019-04-09', end='2024-04-09')

# Download IPC Mexico (^MXX) data from Yahoo Finance
ipc_data = yf.download('^MXX', start='2019-04-09', end='2024-04-09')

# Calculate the growth rates over the past 5 years
sp500_data['yoy_growth'] = (sp500_data['Adj Close']/sp500_data['Close'].shift(252)-1)
sp500_growth = int(sp500_data['yoy_growth'].sum())
ipc_data['yoy_growth'] = (ipc_data['Adj Close']/ipc_data['Close'].shift(252)-1)
ipc_growth = int(ipc_data['yoy_growth'].sum())

# Determine which index had the highest growth rate
better_index = "S&P 500" if sp500_growth > ipc_growth else "IPC Mexico"
higher_growth = round(sp500_growth) if sp500_growth > ipc_growth else round(ipc_growth)

print("The better performing index recently is:", better_index)
print("The growth rate of the better performing index is:", higher_growth, "%")

[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed

The better performing index recently is: S&P 500
The growth rate of the better performing index is: 125 %


Question 4. [Stocks OHLCV] 52-weeks range ratio (2023) for the selected stocks

Find the largest range ratio [=(max-min)/max] of Adj.Close prices in 2023

Download the 2023 daily OHLCV data on Yahoo Finance for top6 stocks on earnings (https://companiesmarketcap.com/most-profitable-companies/): 
2222.SR,BRK-B, AAPL, MSFT, GOOG, JPM.

Here is the example data you should see in Pandas for "2222.SR": https://finance.yahoo.com/quote/2222.SR/history

Calculate maximum-minimim "Adj.Close" price for each stock and divide it by the maximum "Adj.Close" value. 
Round the result to two decimal places (e.g. 0.1575 will be 0.16)

(Additional: why this may be important for your research?)

In [63]:
import yfinance as yf

# Define the ticker symbols for the selected stocks
ticker_symbols = ['2222.SR', 'BRK-B', 'AAPL', 'MSFT', 'GOOG', 'JPM']

# Dictionary to store the maximum and minimum adjusted close prices for each stock
stock_data = {}

# Download 2023 daily OHLCV data for each stock
for symbol in ticker_symbols:
    stock = yf.download(symbol, start='2023-01-01', end='2023-12-31')
    stock_data[symbol] = stock['Adj Close']

# Calculate the maximum and minimum adjusted close prices for each stock
max_min_ratios = {}
for symbol, data in stock_data.items():
    max_price = data.max()
    min_price = data.min()
    max_min_ratio = (max_price - min_price) / max_price
    max_min_ratios[symbol] = round(max_min_ratio, 2)

# Find the stock with the largest range ratio
max_range_stock = max(max_min_ratios, key=max_min_ratios.get)
max_range_ratio = max_min_ratios[max_range_stock]

print("Stock with the largest 52-week range ratio:", max_range_stock)
print("52-week range ratio:", max_range_ratio)

[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed

Stock with the largest 52-week range ratio: MSFT
52-week range ratio: 0.42


Question 5. [Stocks] Dividend Yield

Find the largest dividend yield for the same set of stocks

Use the same list of companies (2222.SR,BRK-B, AAPL, MSFT, GOOG, JPM) and download all dividends paid in 2023. 
You can use get_actions() method or .dividends field in yfinance library (https://github.com/ranaroussi/yfinance?tab=readme-ov-file#quick-start)

Sum up all dividends paid in 2023 per company and divide each value by the closing price (Adj.Close) at the last trading day of the year.

Find the maximum value in % and round to 1 digit after the decimal point. 
(E.g., if you obtained $1.25 dividends paid and the end year stock price is $100, the dividend yield is 1.25% -- and your answer should be equal to 1.3)

In [87]:
import yfinance as yf

# Define the ticker symbols for the selected stocks
ticker_symbols = ['2222.SR', 'BRK-B', 'AAPL', 'MSFT', 'GOOG', 'JPM']

# Dictionary to store the total dividends paid and closing prices for each stock
dividend_yield_data = {}

# Download all dividends paid in 2023 and closing prices for each stock
for symbol in ticker_symbols:
    stock = yf.Ticker(symbol)
    
    # Get dividends paid in 2023
    dividends = stock.dividends.loc['2023']
    total_dividends = dividends.sum()
    
    # Get closing price (Adj. Close) at the last trading day of the year
    try:
        last_trading_day = stock.history(start='2023-12-01', end='2023-12-31')['Close'].iloc[-1]
        
        # Calculate dividend yield
        dividend_yield = (total_dividends / last_trading_day) * 100
        
        # Store dividend yield in the dictionary
        dividend_yield_data[symbol] = dividend_yield
    
    except IndexError:
        print("No data available for {} on 2023-12-31".format(symbol))

# Find the largest dividend yield
max_dividend_yield_symbol = max(dividend_yield_data, key=dividend_yield_data.get)
max_dividend_yield = dividend_yield_data[max_dividend_yield_symbol]

print("Stock with the largest dividend yield:", max_dividend_yield_symbol)
print("Largest dividend yield:", round(max_dividend_yield, 1), "%")


Stock with the largest dividend yield: 2222.SR
Largest dividend yield: 2.8 %


Question 6. [Exploratory] Investigate new metrics

Free text answer

Download and explore a few additional metrics or time series that might be valuable for your project and write down why (briefly).

In [31]:
import yfinance as yf

# Download Nvidia stock data
nvidia = yf.Ticker("NVDA")

# Get historical price data
price_data = nvidia.history(period="1y")

# Get market capitalization
market_cap = nvidia.info['marketCap']

# Get financial metrics
pe_ratio = nvidia.info['forwardPE']
pb_ratio = nvidia.info['priceToBook']
dividend_yield = nvidia.info['dividendYield']
roe = nvidia.info['returnOnEquity']
de_ratio = nvidia.info['debtToEquity']

# Print the metrics
print("Market Capitalization:", market_cap)
print("Price to Earnings Ratio (P/E Ratio):", pe_ratio)
print("Price to Book Ratio (P/B Ratio):", pb_ratio)
print("Dividend Yield:", dividend_yield)
print("Return on Equity (ROE):", roe)
print("Debt to Equity Ratio:", de_ratio)

# Get sector and industry data
sector = nvidia.info['sector']
industry = nvidia.info['industry']

print("Sector:", sector)
print("Industry:", industry)

# Get sentiment analysis data (if available)
# Example: news_sentiment = nvidia.info['sentiment']

Market Capitalization: 1905000054784
Price to Earnings Ratio (P/E Ratio): 24.85323
Price to Book Ratio (P/B Ratio): 43.687653
Dividend Yield: 0.0002
Return on Equity (ROE): 0.91458
Debt to Equity Ratio: 25.725
Sector: Technology
Industry: Semiconductors


In [28]:
import pandas as pd
import yfinance as yf

# Example 1: Stock Price Volatility
def calculate_volatility(symbol):
    # Download historical data
    stock_data = yf.download(symbol, start="2020-01-01", end="2024-01-01")
    # Calculate daily returns
    stock_data['Returns'] = stock_data['Adj Close'].pct_change()
    # Calculate volatility (standard deviation of returns)
    volatility = stock_data['Returns'].std()
    return volatility

# Example usage
volatility = calculate_volatility("NVDA")
print("NVDA Volatility:", volatility)

# Example 2: Trading Volume
def average_trading_volume(symbol):
    # Download historical data
    stock_data = yf.download(symbol, start="2020-01-01", end="2024-01-01")
    # Calculate average trading volume
    avg_volume = stock_data['Volume'].mean()
    return avg_volume

# Example usage
avg_volume = average_trading_volume("NVDA")
print("NVDA Average Trading Volume:", avg_volume)

# Example 3: Moving Averages
def moving_average_crossover(symbol, short_window=50, long_window=200):
    # Download historical data
    stock_data = yf.download(symbol, start="2020-01-01", end="2021-01-01")
    # Calculate short and long moving averages
    stock_data['Short_MA'] = stock_data['Adj Close'].rolling(window=short_window).mean()
    stock_data['Long_MA'] = stock_data['Adj Close'].rolling(window=long_window).mean()
    # Check for crossover
    stock_data['Signal'] = 0
    stock_data.loc[stock_data['Short_MA'] > stock_data['Long_MA'], 'Signal'] = 1
    stock_data.loc[stock_data['Short_MA'] < stock_data['Long_MA'], 'Signal'] = -1
    return stock_data

# Example usage
ma_crossover_data = moving_average_crossover("NVDA")
print(ma_crossover_data[['Adj Close', 'Short_MA', 'Long_MA', 'Signal']].tail())

[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed


NVDA Volatility: 0.034160707503229505
NVDA Average Trading Volume: 46418459.54274354


[*********************100%%**********************]  1 of 1 completed

             Adj Close    Short_MA     Long_MA  Signal
Date                                                  
2020-12-24  129.639221  132.987663  106.436422       1
2020-12-28  128.703857  132.774977  106.779898       1
2020-12-29  129.135345  132.602539  107.180896       1
2020-12-30  131.155731  132.533096  107.565995       1
2020-12-31  130.250320  132.416071  107.964569       1


In [ ]:
# Example 4: Market Breadth Indicators
def market_breadth(symbol):
    # Download historical data of S&P 500 components
    sp500_components = pd.read_csv('https://r2.datahub.io/clt989dvv0003l708h2jgh5ef/main/raw/data/constituents.csv')
    symbols = sp500_components['Symbol'].tolist()
    # Download historical data for each component
    breadth_data = pd.DataFrame()
    for sym in symbols:
        stock_data = yf.download(sym, start="2020-01-01", end="2021-01-01")
        breadth_data[sym] = stock_data['Adj Close']
    # Calculate advancing and declining issues
    breadth_data['Advancing'] = (breadth_data.iloc[-1] > breadth_data.iloc[0]).sum()
    breadth_data['Declining'] = (breadth_data.iloc[-1] < breadth_data.iloc[0]).sum()
    return breadth_data[['Advancing', 'Declining']]

# Example usage
market_breadth_data = market_breadth("^GSPC")
print("Market Breadth:", market_breadth_data)

Question 7. [Exploratory] Time-driven strategy description around earnings releases

Free text answer

Explore earning dates for the whole month of April - e.g. using YahooFinance earnings calendar (https://finance.yahoo.com/calendar/earnings?from=2024-04-21&to=2024-04-27&day=2024-04-23). 
Compare with the previous closed earnings (e.g., recent dates with full data https://finance.yahoo.com/calendar/earnings?from=2024-04-07&to=2024-04-13&day=2024-04-08).

Describe an analytical strategy/idea (you're not required to implement it) to select a subset companies of interest based on the future events data.